File: parser.js
const fs = require("fs");
const request = require("request");
const xmlStream = require("xml-stream");

// NOTE: The default sqlite3 database that is used in Rails
// will curl into a little ball and cry softly to itself if
// it is hit with this script. Be sure to use a database
// engine that properly supports transactions/concurrency
// like PostgreSQL before running this script.

// NOTE 2: For some reason, request makes a distinction
// between 127.0.0.1 and localhost. If utilizing against
// 127.0.0.1, be sure to bind Rails to 127.0.0.1 like so:
// > rails s -b 127.0.0.1

// Change as desired
const COURSECYCLE_API_SERVER = "<SERVER_HERE>";
const FILE_PATH = "<DATA_FILE_HERE>";

// Read file as a stream
const stream = fs.createReadStream(__dirname + FILE_PATH);
const xml = new xmlStream(stream);

// Figure out which elements to retain
xml.preserve("course", true);
xml.on("endElement: course", (course) => {
  buildRequest({
    course: {
      title: course.title.$text,
      description: course.description.$text,
      subject: course.subject.$text,
      code: course.code.$text
    }
  });
});

// Generate a POST request
function buildRequest(form) {
  const coursesUrl = COURSECYCLE_API_SERVER + "/courses";
  request.post({
    headers: {
      "Content-Type": "application/json"
    },
    url: coursesUrl,
    form: form
  }, (err, httpResponse, body) => {
    process.stdout.write(".");
 });
}

File: parser.rb
require "nokogiri"
require "openssl"
require "uri"
require "net/http"

doc = File.open("data/cs107.xml") { |f| Nokogiri::XML(f) }

def persist_course(course)
  url = URI("https://localhost:3000/courses/import")

  http = Net::HTTP.new(url.host, url.port)
  http.use_ssl = true
  http.verify_mode = OpenSSL::SSL::VERIFY_NONE

  request = Net::HTTP::Post.new(url)
  request["content-type"] = 'application/xml'
  request["cache-control"] = 'no-cache'
  request.body = course.to_s

  response = http.request(request)

  print "."
  $stdout.flush
end

courses = doc.xpath("//course")
courses.each do |course|
  persist_course course
end
